domain/storageprovisioning/state/state.go

```diff
@@ -288,3 +315,108 @@ func (st *State) checkApplicationExists(
 
 	return true, nil
 }
+
+// GetStorageIDsForUnit returns the storage IDs for the given unit UUID.
+//
+// The following errors may be returned:
+// - [applicationerrors.UnitNotFound] when no unit exists for the supplied unit UUID.
+// - [corestorage.InvalidStorageID] when the provided unit UUID is invalid.
+func (s *State) GetStorageIDsForUnit(ctx context.Context, unitUUID string) ([]corestorage.ID, error) {
+	db, err := s.DB()
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	input := storageAttachment{
+		UnitUUID: unitUUID,
+	}
+	stmt, err := s.Prepare(`
+SELECT &storageAttachment.*
+FROM   storage_attachment sa
+JOIN   storage_instance si ON sa.storage_instance_uuid = si.uuid
+WHERE  unit_uuid = $storageAttachment.unit_uuid`, input)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	var attachments storageAttachments
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		if exists, err := s.checkUnitExists(ctx, tx, unitUUID); err != nil {
+			return errors.Capture(err)
+		} else if !exists {
+			return errors.Errorf("unit %q does not exist", unitUUID).Add(
+				applicationerrors.UnitNotFound,
+			)
+		}
+
+		err := tx.Query(ctx, stmt, input).GetAll(&attachments)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf("unit %q does not exist", unitUUID).Add(
+				applicationerrors.UnitNotFound,
+			)
+		}
+		return err
+	})
+
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	var storageIDs []corestorage.ID
+	for _, attachment := range attachments {
+		id := corestorage.ID(attachment.StorageID)
+		if err := id.Validate(); err != nil {
+			// This should not happen, but if it does, we return an error.
+			return nil, errors.Errorf(
+				"invalid storage ID %q for unit %q: %w", id, unitUUID, err,
+			).Add(corestorage.InvalidStorageID)
+		}
+		storageIDs = append(storageIDs, id)
+	}
+	return storageIDs, nil
+}
+
+// GetAttachmentLife retrieves the life of a storage attachment for a unit.
+//
+// The following errors may be returned:
+// - [applicationerrors.UnitNotFound] when no unit exists for the supplied unit UUID.
+// - [storageprovisioningerrors.AttachmentNotFound] when the storage attachment does not exist for the unit.
+func (s *State) GetAttachmentLife(ctx context.Context, unitUUID, storageID string) (domainlife.Life, error) {
+	db, err := s.DB()
+	if err != nil {
+		return -1, errors.Capture(err)
+	}
+	attachment := storageAttachmentLife{
+		UnitUUID:  unitUUID,
+		StorageID: storageID,
+	}
+	stmt, err := s.Prepare(`
+SELECT (si.life_id) AS (&storageAttachmentLife.*)
+FROM   storage_attachment sa
+JOIN   storage_instance si ON sa.storage_instance_uuid = si.uuid
+WHERE  sa.unit_uuid = $storageAttachmentLife.unit_uuid
+AND    si.storage_id = $storageAttachmentLife.storage_id`, attachment)
+	if err != nil {
+		return -1, errors.Capture(err)
+	}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		if exists, err := s.checkUnitExists(ctx, tx, unitUUID); err != nil {
```

Comment:
how do we know if the storage instance  exists ?  We also need to check that this exists.